src/rbd-use.cc

```diff
@@ -0,0 +1,663 @@
+  if (r < 0)
+    goto close_nbd;
+
+  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
+    r = -errno;
+    goto close_nbd;
+  }
+
+  r = ioctl(nbd, NBD_SET_SOCK, fd[0]);
+  if (r < 0)
+    goto close_fd;
+
+  memset(&sa, 0, sizeof(sa));
+  sa.sa_handler = SIG_IGN;
+  sigaction(SIGCHLD, &sa, NULL);
+
+  cout << devpath << std::endl;
+
+  if (!foreground) {
+    r = daemonhelper.daemon();
+    if (r < 0)
+      goto close_fd;
+  }
+
+  {
+    NBDServer server(fd[1], image);
+
+    server.start();
+    ioctl(nbd, NBD_DO_IT);
+    server.stop();
+  }
+
+close_fd:
+  close(fd[0]);
+  close(fd[1]);
+close_nbd:
+  close(nbd);
+  image.close();
+  io_ctx.close();
+  rados.shutdown();
+  return r;
+}
+
+static int do_unmap()
+{
+  int nbd = open(devpath.c_str(), O_RDWR);
+  if (nbd < 0)
+    return nbd;
+
+  int r = ioctl(nbd, NBD_DISCONNECT);
+  if (r < 0)
+    goto out;
+  r = ioctl(nbd, NBD_CLEAR_SOCK);
+
+out:
+  close(nbd);
+  return r;
+}
+
+static void parse_imgpath(const char *imgpath)
+{
+  if (!imgpath)
+    return;
+
+  const char *pos = imgpath;
+  const char *start = imgpath;
+  while (*pos != 0) {
+    if (*pos == '/') {
+      poolname.assign(start, pos - start);
+      start = pos + 1;
+    } else if (*pos == '@') {
+      imgname.assign(start, pos - start);
+      start = pos + 1;
+    }
+    ++pos;
+  }
+  if (poolname.empty())
+    poolname = "rbd";
+  if (imgname.empty())
+    imgname.assign(start, pos - start);
+  else
+    snapname.assign(start, pos - start);
+}
+
+static bool find_empty_nbd(std::string &devpath)
+{
+  int id = 0;
+  int try_load = 0;
+  char path[64];
+
+  devpath.clear();
+  while (devpath.empty()) {
+    snprintf(path, sizeof(path), "/dev/nbd%d", id);
+    int nbd = open(path, O_RDWR);
+    if (nbd < 0) {
+      if (!try_load) {
+	int r = module_load("nbd", NULL);
+	if (r < 0) {
+		cerr << "rbd-use: failed to load nbd kernel module: " << cpp_strerror(-r) << std::endl;
+		break;
+	}
+	try_load = 1;
+	continue;
+      }
+      cerr << "rbd-use: failed to open nbd device: " << cpp_strerror(-nbd) << std::endl;
+      break;
+    }
+
+    int r = ioctl(nbd, NBD_DO_IT);
+    close(nbd);
+    if (r < 0) {
+	if (errno == EINVAL) {
+		devpath = path;
+		break;
+	}
+	++id;
+    } else {
+	break;
+    }
+  }
+
+  return !devpath.empty();
+}
+
+static void list_mapped_devices()
+{
+  char path[64];
+  int m = 0;
+
+  while (1) {
+    snprintf(path, sizeof(path), "/dev/nbd%d", m);
+    int nbd = open(path, O_RDWR);
+    if (nbd < 0)
+      break;
+    int r = ioctl(nbd, NBD_DO_IT);
+    if (r < 0 && errno == EBUSY)
+      cout << path << std::endl;
+    close(nbd);
+    m++;
+  }
+}
+
+int main(int argc, const char *argv[])
+{
+  int r;
+  vector<const char*> args;
+
+  r = daemonhelper.init();
+  if (r < 0)
+    return EXIT_FAILURE;
+
+  argv_to_vec(argc, argv, args);
+  env_to_vec(args);
+
+  global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT, CODE_ENVIRONMENT_UTILITY, 0);
```

Comment:
like ceph-fuse, it seems to make sense to use CODE_ENVIRONMENT_DAEMON and CINIT_FLAG_UNPRIVILEGED_DAEMON_DEFAULTS here